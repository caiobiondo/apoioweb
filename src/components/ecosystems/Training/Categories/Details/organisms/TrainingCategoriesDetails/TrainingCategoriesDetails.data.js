import gql from 'graphql-tag';
import { APP_VERSION, Origem } from '../../../../../../../config';

const ITEMS_PER_PAGE = 20;

export const TrainingCategoriesDetailsQuery = gql`
  query TrainingCategoriesDetails(
    $sellerId: Int!
    $categoryId: Int!
    $offset: Int!
    $limit: Int!
    $ciclo: Int
    $setor: Int
    $gerenciaMercado: Int
    $grupo: Int
    $papelDaConsultora: Int
    $canal: Int
    $gerenciaDeVendas: Int
    $regiao: Int
    $appVersion: String!
    $origem: String!
  ) {
    trainingCoursesByCategory(
      sellerId: $sellerId
      categoryId: $categoryId
      offset: $offset
      limit: $limit
      ciclo: $ciclo
      setor: $setor
      gerenciaMercado: $gerenciaMercado
      grupo: $grupo
      papelDaConsultora: $papelDaConsultora
      canal: $canal
      gerenciaDeVendas: $gerenciaDeVendas
      regiao: $regiao
      appVersion: $appVersion
      origem: $origem
    ) {
      hasNextPage
      items {
        id
        title
        thumbnail
        dateUpload
        durationInSeconds
        type
        views
        stoppedAt
        status
        isfavorite
      }
    }
    trainingCategory(
      sellerId: $sellerId
      categoryId: $categoryId
      ciclo: $ciclo
      setor: $setor
      gerenciaMercado: $gerenciaMercado
      grupo: $grupo
      papelDaConsultora: $papelDaConsultora
      canal: $canal
      gerenciaDeVendas: $gerenciaDeVendas
      regiao: $regiao
      appVersion: $appVersion
      origem: $origem
    ) {
      id
      name
      thumbnail
      banners {
        id
        thumbnail
        categoryId
      }
      totalOfCourses
      totalOfCoursesCompleted
    }
  }
`;

export const updateQuery = (previousResult, { fetchMoreResult }) => {
  if (!fetchMoreResult) {
    return previousResult;
  }

  const previousResultIds = previousResult.trainingCoursesByCategory.items.map(item => item.id);
  const fetchMoreResultsToAdd = fetchMoreResult.trainingCoursesByCategory.items.filter(
    item => previousResultIds.indexOf(item.id) < 0,
  );

  return Object.assign({}, previousResult, {
    trainingCoursesByCategory: {
      __typename: previousResult.trainingCoursesByCategory.__typename,
      hasNextPage: fetchMoreResult.trainingCoursesByCategory.hasNextPage,
      items: [...previousResult.trainingCoursesByCategory.items, ...fetchMoreResultsToAdd],
    },
  });
};

export const TrainingCategoriesDetailsOptions = {
  options(props) {
    return {
      variables: {
        sellerId: props.user.codigo,
        categoryId: props.categoryId,
        limit: ITEMS_PER_PAGE,
        offset: 0,
        ciclo:
          props.user.estrutura.ciclo &&
          props.user.estrutura.ciclo[0] &&
          props.user.estrutura.ciclo[0].numero,
        setor: props.user.estrutura.codigoTipo > 3 ? props.user.estrutura.setor.codigo : 0,
        gerenciaMercado: props.user.estrutura.gerenciaMercado.codigo,
        grupo: props.user.estrutura.codigoTipo > 4 ? props.user.estrutura.codigo : 0,
        papelDaConsultora: props.user.cdPapelAtivo,
        canal: props.user.cdCanalCaptacao,
        gerenciaDeVendas:
          props.user.estrutura.codigoTipo > 2 ? props.user.estrutura.gerenciaVenda.codigo : 0,
        regiao:
          props.user.estrutura.codigoTipo > 1 ? props.user.estrutura.regiaoEstrategica.codigo : 0,
        appVersion: APP_VERSION,
        origem: Origem,
      },
      fetchPolicy: 'cache-and-network',
    };
  },
  props({ data }) {
    const { refetch, loading, trainingCategory } = data;
    const trainingCourses = data.trainingCoursesByCategory && data.trainingCoursesByCategory.items;
    const hasNextPage =
      (data.trainingCoursesByCategory && data.trainingCoursesByCategory.hasNextPage) || false;

    return {
      data,
      refetch: refetch,
      loading: loading,
      trainingCourses,
      trainingCategory,
      hasNextPage,
      fetchMore() {
        if (data.loading) {
          return;
        }

        const offset = trainingCourses ? trainingCourses.length : 0;
        const variables = { offset };

        return data.fetchMore({
          variables,
          updateQuery,
        });
      },
    };
  },
};
