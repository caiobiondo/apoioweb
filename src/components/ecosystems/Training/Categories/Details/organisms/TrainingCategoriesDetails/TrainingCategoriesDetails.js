import React, { Component } from 'react';
import EmptyList from 'components/molecules/EmptyList/EmptyList';
import {
  TrainingCategoriesDetailsQuery,
  TrainingCategoriesDetailsOptions,
} from './TrainingCategoriesDetails.data';
import { TrainingCourseUpdateMutation } from 'components/ecosystems/Training/data/TrainingCourseUpdate.data';
import { graphql, compose } from 'react-apollo';

import TrainingCategoriesDetailsHeader from '../../molecules/Header/TrainingCategoriesDetailsHeader';
import TrainingCourses from 'components/ecosystems/Training/molecules/TrainingCourses';
import InfiniteScroll from 'components/organisms/InfiniteScroll';

import { translate } from 'locale';
import MenuItem from 'material-ui/MenuItem';
import IconMenu from 'material-ui/IconMenu';
import IconButton from 'material-ui/IconButton';
import MoreVertIcon from 'material-ui/svg-icons/navigation/more-vert';
import Dialog from 'material-ui/Dialog';
import FlatButton from 'material-ui/FlatButton';
import { RobotoRegular } from 'styles/typography';

import { injectIntl, FormattedMessage } from 'react-intl';
import { getHeadersFromUser } from '../../../../../../../utils/getUserParams';

import {
  TrainingCategoriesDetailsWrapper,
  TrainingCategoriesDetailsContentWrapper,
  TitleWrapper,
  CategoryIcon,
  Title,
  TrainingCourseFeedbackModalTitle,
  TrainingCourseFeedbackModalAction,
} from './TrainingCategoriesDetails.styles';

export class TrainingCategoriesDetails extends Component {
  constructor(props) {
    super(props);

    this.state = {
      feedbackModalOpened: false,
      feedbackModalTitle: '',
    };
  }

  componentWillReceiveProps({ loading, trainingCourses }) {
    this.notifyLoadFinish(loading, trainingCourses);
  }

  notifyLoadFinish = (loading, items) => {
    if (!loading && this.props.onLoadFinished) {
      this.props.onLoadFinished(this.isEmpty(loading, items), this.isLoading(loading, items));
    }
  };

  isLoading = (loading, items) => {
    return loading && !items;
  };

  isEmpty = (loading, items) => {
    return !loading && (!items || items.length === 0);
  };

  handleMenuItemClick = (event, child) => {
    const { formatMessage } = this.props.intl;
    const {
      ciclo,
      grupo,
      gerenciaDeVendas,
      regiao,
      setor,
      gerenciaMercado,
      papelDaConsultora,
      canal,
      appVersion,
      origem,
    } = getHeadersFromUser(this.props.user);
    this.props
      .mutate({
        variables: {
          input: { action: child.props.value },
          sellerId: this.props.user.codigo,
          courseId: child.props.course.id,
          ciclo,
          grupo,
          gerenciaDeVendas,
          regiao,
          setor,
          gerenciaMercado,
          papelDaConsultora,
          canal,
          appVersion,
          origem,
        },
      })
      .then(response => {
        if (response.error) {
          const message =
            child.props.value === 'favorite'
              ? formatMessage({ id: 'trainingAddCourseError' })
              : formatMessage({ id: 'trainingRemoveCourseError' });
          this.setState({
            feedbackModalOpened: true,
            feedbackModalTitle: message,
          });

          // Handle error
          return;
        }

        if (!response.data.updateCourse.status) {
          const message =
            child.props.value === 'favorite'
              ? formatMessage({ id: 'trainingAddCourseFailure' })
              : formatMessage({ id: 'trainingRemoveCourseFailure' });
          this.setState({
            feedbackModalOpened: true,
            feedbackModalTitle: message,
          });
          // Handle not updated
          return;
        }

        // Handle update success
        this.props.refetch();
        const message =
          child.props.value === 'favorite'
            ? formatMessage({ id: 'trainingAddCourseSuccess' })
            : formatMessage({ id: 'trainingRemoveCourseSuccess' });

        this.setState({
          feedbackModalOpened: true,
          feedbackModalTitle: message,
        });

        return;
      })
      .catch(err => {
        console.log('err', err);
        // Handle error
        const message =
          child.props.value === 'favorite'
            ? formatMessage({ id: 'trainingAddCourseError' })
            : formatMessage({ id: 'trainingRemoveCourseError' });
        this.setState({
          feedbackModalOpened: true,
          feedbackModalTitle: message,
        });
      });
  };

  renderMenuItems = course => {
    const style = { fontFamily: RobotoRegular };

    if (course.isfavorite === 'true') {
      return (
        <IconMenu
          iconButtonElement={
            <IconButton>
              <MoreVertIcon />
            </IconButton>
          }
          anchorOrigin={{ horizontal: 'right', vertical: 'top' }}
          targetOrigin={{ horizontal: 'right', vertical: 'top' }}
          onItemTouchTap={this.handleMenuItemClick}
        >
          <MenuItem
            style={style}
            primaryText={translate('trainingRemoveCourseMyList')}
            value="unfavorite"
            course={course}
          />
        </IconMenu>
      );
    }

    return (
      <IconMenu
        iconButtonElement={
          <IconButton>
            <MoreVertIcon />
          </IconButton>
        }
        anchorOrigin={{ horizontal: 'right', vertical: 'top' }}
        targetOrigin={{ horizontal: 'right', vertical: 'top' }}
        onItemTouchTap={this.handleMenuItemClick}
      >
        <MenuItem
          style={style}
          primaryText={translate('trainingAddCourseMyList')}
          value="favorite"
          course={course}
        />
      </IconMenu>
    );
  };

  handleClose = () => {
    this.setState({ feedbackModalOpened: false });
  };

  renderFeedbackModal = () => {
    const { feedbackModalTitle } = this.state;
    const actions = [
      <FlatButton
        label={<FormattedMessage id="ok" />}
        primary={true}
        labelStyle={TrainingCourseFeedbackModalAction}
        onClick={this.handleClose}
      />,
    ];

    return (
      <Dialog
        key="feedbackModal"
        title={feedbackModalTitle}
        actions={actions}
        modal={false}
        open={this.state.feedbackModalOpened}
        titleStyle={TrainingCourseFeedbackModalTitle}
        onRequestClose={this.handleClose}
      />
    );
  };

  render() {
    const { loading, hasNextPage, trainingCourses, trainingCategory, fetchMore } = this.props;

    if (!trainingCategory) {
      return null;
    }

    return (
      <TrainingCategoriesDetailsWrapper>
        <TrainingCategoriesDetailsHeader category={trainingCategory} />
        <TrainingCategoriesDetailsContentWrapper>
          <TitleWrapper>
            <CategoryIcon src={trainingCategory.thumbnail} alt={trainingCategory.name} />
            <Title>{trainingCategory.name}</Title>
          </TitleWrapper>
          <InfiniteScroll
            onScroll={fetchMore}
            hasMore={hasNextPage}
            loading={loading}
            debounce={500}
            items={trainingCourses}
            emptyList={
              <EmptyList
                icon="ico_list_add"
                titleId="coursesEmptyList"
                descriptionId="coursesEmptyListDescription"
              />
            }
          >
            <TrainingCourses courses={trainingCourses} renderMenuItems={this.renderMenuItems} />
          </InfiniteScroll>
          {this.renderFeedbackModal()}
        </TrainingCategoriesDetailsContentWrapper>
      </TrainingCategoriesDetailsWrapper>
    );
  }
}

export const TrainingCategoriesDetailsWithIntl = injectIntl(TrainingCategoriesDetails);

export default compose(
  graphql(TrainingCategoriesDetailsQuery, TrainingCategoriesDetailsOptions),
  graphql(TrainingCourseUpdateMutation),
)(TrainingCategoriesDetailsWithIntl);
