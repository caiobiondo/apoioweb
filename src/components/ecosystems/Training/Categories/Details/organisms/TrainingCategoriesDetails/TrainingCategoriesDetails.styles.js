import styled from 'styled-components';
import { LoadingWrapperStyle } from 'styles/mixins';
import { fs18, fw300 } from 'styles/typography';

import { spMedium, spPage } from 'styles/spacing';

export const LoadingWrapper = LoadingWrapperStyle;

export const fullContainer = `
  height: '100%',
`;

export const TrainingCategoriesDetailsWrapper = styled.div`
  position: relative;
`;

export const TrainingCategoriesDetailsContentWrapper = styled.div`
  flex: 1 1 0;
  flex-direction: column;
  justify-content: center;
  padding-bottom: ${spPage};
  box-shadow: 0px 2px 5px 0 rgba(0, 0, 0, 0.2);
  background-color: white;
`;

export const TitleWrapper = styled.div`
  padding: 25px;
  position: relative;
`;

export const CategoryIcon = styled.img`
  left: 25px;
  position: absolute;
  top: 50%;
  transform: translate(0, -50%);
  width: 20px;
`;

export const Title = styled.h3`
  display: inline-block;
  margin: 0;
  padding: 0 20px 0 40px;
  font-size: ${fs18};
  font-weight: ${fw300};
`;

export const List = styled.ul`
  padding-left: 0;
  padding-top: ${spMedium};
  margin: ${spMedium};
  list-style-type: none;
`;

export const ListItem = styled.li`
  list-style-type: none;
  margin-bottom: 70px;
`;
