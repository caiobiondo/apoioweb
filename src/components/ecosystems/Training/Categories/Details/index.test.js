import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { shallow } from 'enzyme';
import TrainingCategoriesDetails from './index';

const baseState = {
  empty: false,
  loading: true,
};

describe('TrainingCategoriesDetails Ecosystem', () => {
  it('should render the training courses list page', () => {
    const renderer = new ShallowRenderer();
    const params = { id: 1 };

    renderer.render(<TrainingCategoriesDetails match={{ params }} />);
    const result = renderer.getRenderOutput();

    expect(result).toMatchSnapshot();
  });

  it('should update empty and loading state', () => {
    const expectedState = { ...baseState, empty: true, loading: false };
    const params = { id: 1 };

    const result = shallow(<TrainingCategoriesDetails match={{ params }} />);
    const instance = result.instance();
    instance.onLoadFinished(true, false);

    expect(instance.state).toEqual(expectedState);
  });
});
