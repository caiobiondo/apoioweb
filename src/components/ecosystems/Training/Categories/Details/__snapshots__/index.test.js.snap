// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TrainingCategoriesDetails Ecosystem should render the training courses list page 1`] = `
<indexstyles__Main>
  <Apollo(TrainingCategoriesDetails)
    categoryId={1}
    onLoadFinished={[Function]}
    user={undefined}
  />
</indexstyles__Main>
`;
